cns/fakes/cnsfake.go

```diff
@@ -2,17 +2,143 @@ package fakes
 
 import (
 	"encoding/json"
+	"errors"
+	"log"
+	"net"
+	"sync"
 
 	"github.com/Azure/azure-container-networking/cns"
 	"github.com/Azure/azure-container-networking/cns/common"
 )
 
+// available IP's stack
+// all IP's map
+
+type StringStack struct {
+	lock  sync.Mutex // you don't have to do this if you don't want thread safety
+	items []string
+}
+
+func NewStack() *StringStack {
+	return &StringStack{sync.Mutex{}, make([]string, 0)}
+}
+
+func (stack *StringStack) Push(v string) {
+	stack.lock.Lock()
+	defer stack.lock.Unlock()
+
+	stack.items = append(stack.items, v)
+}
+
+func (stack *StringStack) Pop() (string, error) {
+	stack.lock.Lock()
+	defer stack.lock.Unlock()
+
+	length := len(stack.items)
+	if length == 0 {
+		return "", errors.New("Empty Stack")
+	}
+
+	res := stack.items[length-1]
+	stack.items = stack.items[:length-1]
+	return res, nil
+}
+
+func incrementIP(ip net.IP) {
+	for j := len(ip) - 1; j >= 0; j-- {
+		ip[j]++
+		if ip[j] > 0 {
+			break
+		}
+	}
+}
+
+type IPStateManager struct {
+	AvailableIPConfigState      map[string]cns.IPConfigurationStatus
+	AllocatedIPConfigState      map[string]cns.IPConfigurationStatus
+	PendingReleaseIPConfigState map[string]cns.IPConfigurationStatus
+	AvailableIPIDStack          StringStack
```

Comment:
Add an `AllocatedIPConfigID map[json.OrchestratorContext]string`, for constant time access when pods are called again with same orchestrator context (pod restart or node restart)